src/shared/copy.c

```diff
@@ -202,6 +202,47 @@ int copy_bytes_full(
                 if (max_bytes != UINT64_MAX && m > max_bytes)
                         m = max_bytes;
 
+                if (copy_flags & COPY_HOLES) {
+                        off_t c, e;
+
+                        c = lseek(fdf, 0, SEEK_CUR);
+                        if (c < 0)
+                                return -errno;
+
+                        /* To see if we're in a hole, we search for the next data offset. */
+                        e = lseek(fdf, c, SEEK_DATA);
+                        if (e < 0 && errno == ENXIO) {
+                                /* If errno == ENXIO, that means we've reached the final hole of the file and
+                                * that hole isn't followed by more data. */
+                                e = lseek(fdf, 0, SEEK_END);
+                                if (e < 0)
+                                        return -errno;
+                        } else if (e < 0)
+                                return -errno;
+
+                        /* If we're in a hole (current offset is not a data offset), create a hole of the same size
+                        * in the target file. */
+                        if (e > c && lseek(fdt, e - c, SEEK_CUR) < 0)
```

Comment:
Hmm, I am not sure how I feel about this. The thing is that this just skips over the target bytes, regardless what they currently are. This works OK if the holes are in the middle of the files *and* the target file is empty so far. But this will go wrong if the hole is at the end of the file, because no write() or ftruncate() is done, so the seek will have zero effect...

Not sure what the best way to fix this is.

At the very least we need to a new flags COPY_TRUNCATE or so, that causes a truncation of the target file once the copy operation is complete, and most calls of COPY_HOLES also need to specify that (except if they operate piecemeal, and call ftruncate on their own in the end).

We might also use fallocate() with FALLOC_FL_PUNCH_HOLE to punch the hole explicitly. I think this would be much nicer in most cases, since regardless what the target fd points to currently the end result would be correct. i.e. we'd always combine FALLOC_FL_PUNCH_HOLE with lseek(). Doing this requires some careful coding to ensure we have good fallbacks, since FALLOC_FL_PUNCH_HOLE is only supported on some file systems and we should try to make sure we don't need more than two syscalls for this, i.e. we need to correctly track the offset to pass to fallocate().

Reply:
See also https://github.com/systemd/systemd/pull/22250 which addresses some of these concerns but not all. I was assuming FALLOC_FL_PUNCH_HOLE was supported everywhere.